src/lib/os-release.c

```diff
@@ -0,0 +1,235 @@
+                }
+
+                cur = (l - buf);
+                r -= cur;
+
+                /* Look for assignment */
+                c = memchr(l, '=', (size_t)r);
+                if (!c) {
+                        goto next_line;
+                }
+
+                /* Skip empty value */
+                val_len = (size_t)(r - ((c - l) + 1));
+                if (val_len < 1) {
+                        goto next_line;
+                }
+
+                key = strndup(l, (size_t)(c - l));
+                value = strndup(c + 1, val_len);
+
+                /* Fix the key to always be upper case */
+                char *tkey = key;
+                while (*tkey != '\0') {
+                        *tkey = (char)toupper(*tkey);
+                        ++tkey;
+                }
+
+                if (value[val_len - 1] == '\'' || value[val_len - 1] == '\"') {
+                        value[val_len - 1] = '\0';
+                }
+
+                if (value[0] == '\'' || value[0] == '\"') {
+                        incr = 1;
+                }
+
+                if (!nc_hashmap_put(self, key, strdup(value + incr))) {
+                        ret = false;
+                        break;
+                }
+        next_line:
+                free(buf);
+                buf = NULL;
+        }
+
+        if (buf) {
+                free(buf);
+                buf = NULL;
+        }
+        return ret;
+}
+
+CbmOsRelease *cbm_os_release_new(const char *path)
+{
+        NcHashmap *ret = NULL;
+
+        ret = nc_hashmap_new_full(nc_string_hash, nc_string_compare, free, free);
+        if (!cbm_os_release_parse(ret, path)) {
+                cbm_os_release_free(ret);
+                return nc_hashmap_new_full(nc_string_hash, nc_string_compare, free, free);
+        }
+
+        return ret;
+}
+
+CbmOsRelease *cbm_os_release_new_for_root(const char *root)
+{
+        static const char *files[] = { "etc/os-release", "usr/lib/os-release" };
+        CbmOsRelease *release = NULL;
+
+        /* Loop the files until one parses, otherwise return an empty NcHashmap */
+        for (size_t i = 0; i < ARRAY_SIZE(files); i++) {
+                autofree(char) *p = NULL;
+
+                if (asprintf(&p, "%s/%s", root, files[i]) < 0) {
+                        DECLARE_OOM();
+                        return NULL;
+                }
+
+                if (!nc_file_exists(p)) {
+                        continue;
+                }
+
+                release = cbm_os_release_new(p);
+                if (release) {
+                        return release;
+                }
+        }
+
+        return nc_hashmap_new_full(nc_string_hash, nc_string_compare, free, free);
+}
+
+void cbm_os_release_free(CbmOsRelease *self)
+{
+        nc_hashmap_free(self);
+}
+
+const char *cbm_os_release_get_value(CbmOsRelease *self, CbmOsReleaseKey key)
+{
+        const char *strkey = NULL;
+        const char *ret = NULL;
+
+        if (!self) {
```

Comment:
So if a nc_hashmap_new fails due to malloc in theory get could be called on null (since new_for_root just returns without checking to see if nc_hashmap_new was successful). In that case do you want the returned value be cbm_os_release_fallback_value instead of NULL?